src/compute/src/render/reduce.rs

```diff
@@ -555,33 +643,76 @@ where
+        let mut datums1 = DatumVec::new();
+        let mut datums2 = DatumVec::new();
+        let mfp_after1 = mfp_after.clone();
+        let mfp_after2 = mfp_after;
+
+        let arranged =
+            differential_dataflow::collection::concatenate(&mut input.scope(), to_collect)
+                .mz_arrange::<RowValSpine<_, _, _>>("Arranged ReduceFuseBasic input");
+
+        let output = arranged.mz_reduce_abelian::<_, RowRowSpine<_, _>>("ReduceFuseBasic", {
+            move |key, input, output| {
+                let temp_storage = RowArena::new();
+                let datum_iter = key.into_datum_iter(None);
+                let mut datums_local = datums1.borrow();
+                datums_local.extend(datum_iter);
+                let key_len = datums_local.len();
+
+                for ((_, row), _) in input.iter() {
+                    datums_local.push(row.unpack_first());
                 }
-            });
-        (
-            output,
-            err_output.expect("expected to validate in at least one aggregate"),
-        )
+
+                if let Some(row) =
+                    evaluate_mfp_after(&mfp_after1, &mut datums_local, &temp_storage, key_len)
+                {
+                    output.push((row, 1));
+                }
+            }
+        });
+        // If `mfp_after` can error, then we need to render a paired reduction
+        // to scan for these potential errors. Note that we cannot directly use
+        // `mz_timely_util::reduce::ReduceExt::reduce_pair` here because we only
+        // conditionally render the second component of the reduction pair.
+        let validation_errs = err_output.expect("expected to validate in at least one aggregate");
+        if let Some(mfp) = mfp_after2.filter(|mfp| mfp.could_error()) {
+            let mfp_errs = arranged
+                .mz_reduce_abelian::<_, RowErrSpine<_, _>>(
+                    "ReduceFuseBasic Error Check",
+                    move |key, input, output| {
+                        // Since negative accumulations are checked in at least one component
+                        // aggregate, we only need to look for MFP errors here.
+                        let temp_storage = RowArena::new();
+                        let datum_iter = key.into_datum_iter(None);
+                        let mut datums_local = datums2.borrow();
+                        datums_local.extend(datum_iter);
+
+                        for ((_, row), _) in input.iter() {
+                            datums_local.push(row.unpack_first());
+                        }
+
+                        if let Result::Err(e) = mfp.evaluate_inner(&mut datums_local, &temp_storage)
+                        {
+                            output.push((e.into(), 1));
+                        }
+                    },
+                )
+                .as_collection(|_, v| v.into_owned());
+            (output, validation_errs.concat(&mfp_errs))
```

Comment:
This block is uncovered, seems quite interesting. Can we test it explicitly?

Reply:
Thanks for the coverage run! I think that you mean the `ReduceMonotonic Error Check` block, right? Many of the parts of the code are very similar. I added a commit with extra tests now that should cover this larger uncovered block (by running a one-shot SELECT), and also cover the following by including a basic aggregation with `DISTINCT`:
```diff
@@ -867,7 +852,6 @@ where
                 )
                 .as_collection(|_, v| v.into_owned());
             if let Some(e) = err_output {
!                err_output = Some(e.concat(&errs));
             } else {
                 err_output = Some(errs);
             }
```